zookeeper-server/src/main/java/org/apache/zookeeper/server/auth/znode/groupacl/ZkClientUriDomainMappingHelper.java

```diff
@@ -121,15 +137,43 @@ private void parseZNodeMapping() {
           "ZkClientUriDomainMappingHelper::parseZNodeMapping(): No application domain ZNodes found in root path: {}",
           rootPath);
     }
+    clientUriToDomainNames = newClientUriToDomainNames;
   }
 
   @Override
   public void process(WatchedEvent event) {
     parseZNodeMapping();
+    // Update AuthInfo for all the known connections.
+    ServerCnxnFactory factory =
+        zks.getSecureServerCnxnFactory() == null ? zks.getServerCnxnFactory() : zks.getSecureServerCnxnFactory();
+    factory.getConnections().forEach(cnxn -> updateDomainBasedAuthInfo(cnxn));
   }
 
   @Override
   public Set<String> getDomains(String clientUri) {
     return clientUriToDomainNames.getOrDefault(clientUri, Collections.emptySet());
   }
+
+  @Override
+  public void updateDomainBasedAuthInfo(ServerCnxn cnxn) {
+    if (updater != null && cnxn != null) {
+      // UpdateAuthInfo is triggered on new connection, as well as new domain information.
+      // To prevent inconsistent update, concurrency control is necessary.
+      synchronized (updater) {
+        updater.updateAuthInfo(cnxn, clientUriToDomainNames);
+      }
+    }
+  }
+
+  /**
+   * Interface that declares method(s) to update connection AuthInfo when the client URI to domain mapping is updated.
+   */
+  interface ConnectionAuthInfoUpdater {
```

Comment:
I would prefer to separate it when we have the HttpClientUriDomainMappingHelper and it turns out we will need this interface.

It is very possible that the other implementation auto-updates the AuthInfo in some other ways so this interface is not necessary.

Reply:
Discussed this offline. We agreed to make this a standalone interface because 1) there could be other implementations of the uri domain mapping helper (other than the ZKclient based one), 2) we want to let the implementations to decide whether this type of updater is necessary (we want to make it optional - this is why we are not putting this in `ClientUriDomainMappingHelper` interface).